server/src/com/thoughtworks/go/config/CachedGoConfig.java

```diff
@@ -18,36 +18,207 @@
-import com.thoughtworks.go.server.service.PipelineConfigService;
-
-public interface CachedGoConfig {
-    CruiseConfig loadForEditing();
-
-    CruiseConfig currentConfig();
-
-    void forceReload();
-
-    void loadConfigIfNull();
-
-    ConfigSaveState writeWithLock(UpdateConfigCommand updateConfigCommand);
-
-    EntityConfigSaveResult writeEntityWithLock(EntityConfigUpdateCommand updateConfigCommand, Username currentUser);
-
-    String getFileLocation();
-
-    void save(String configFileContent, boolean shouldMigrate) throws Exception;
-
-    GoConfigValidity checkConfigFileValid();
-
-    void registerListener(ConfigChangedListener listener);
-
-    void clearListeners();
-
-    void reloadListeners();
-
-    GoConfigHolder loadConfigHolder();
-
-    boolean hasListener(ConfigChangedListener listener);
+import com.thoughtworks.go.serverhealth.HealthStateType;
+import com.thoughtworks.go.serverhealth.ServerHealthService;
+import com.thoughtworks.go.serverhealth.ServerHealthState;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.thoughtworks.go.server.service.GoConfigService.INVALID_CRUISE_CONFIG_XML;
+import static com.thoughtworks.go.util.ExceptionUtils.bomb;
+
+/**
+ * @understands when to reload the config file or other config source
+ */
+@Component
+public class CachedGoConfig {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CachedGoConfig.class);
+    private final GoFileConfigDataSource dataSource;
+    private final ServerHealthService serverHealthService;
+    private List<ConfigChangedListener> listeners = new ArrayList<>();
+    private volatile CruiseConfig currentConfig;
+    private volatile CruiseConfig currentConfigForEdit;
+    private volatile GoConfigHolder configHolder;
+    private volatile Exception lastException;
+
+    @Autowired
+    public CachedGoConfig(ServerHealthService serverHealthService,
+                          GoFileConfigDataSource dataSource) {
+        this.serverHealthService = serverHealthService;
+        this.dataSource = dataSource;
+    }
+
+    public static List<ConfigErrors> validate(CruiseConfig config) {
+        List<ConfigErrors> validationErrors = new ArrayList<ConfigErrors>();
+        validationErrors.addAll(config.validateAfterPreprocess());
+        return validationErrors;
+    }
+
+    //used in tests
+    public void throwExceptionIfExists() {
+        if (lastException != null) {
+            throw bomb("Invalid config file", lastException);
+        }
+    }
+
+    //NOTE: This method is called on a thread from Spring
+    public void onTimer() {
+        this.forceReload();
+    }
+
+    public void forceReload() {
+        LOGGER.debug("Config file (on disk) update check is in queue");
+        synchronized (GoConfigWriteLock.class) {
```

Comment:
yes, its not a part of this PR, intentionally else you would never merge my PR :). I can make that change and send another PR for it. I think if [this test](https://github.com/gocd/gocd/blob/master/server/test/integration/com/thoughtworks/go/server/service/ConfigSaveDeadlockDetectionIntegrationTest.java) passes, we are good. What do you think?